### PR TITLE
CAS-60 bedspace end date

### PR DESCRIPTION
**Pull request by froddd**
# Context

https://dsdmoj.atlassian.net/browse/CAS-60

Availability reporting for bedspaces suffers from the overuse of 'voids' to makr bedspace when they are no longer available. Those 'voids' are typically used when the bedspace is not available for a finite period of time; during this time the room is still classed as 'available' in reporting. 

To enable users to archive bedspace when they are no longer available at all, and therefore provide more accurate reporting of availability, they need to be able to set an 'end date' for the bedspace. In practice, this means:

- Once a bedspace has reached its end date, it is automatically archived and cannot be booked
- A bedspace can be created with or without an end date
- An end date can be added when editing a bedspace. Once added, the bedspace end date can no longer be edited
- A bedspace end date cannot:
  - conflict with an existing booking
  - be set to before the creation date of the bedspace

The bedpsace status is then derived from the set end date and shown in various places around the service. The status shows as 'Online' if the bedspace end date is not set or set in the future, and shows as 'Archived' if it is set today or in the past.

# Changes in this PR

### After
Screenshot 2024-03-27 at 18 50 39

Screenshot 2024-03-27 at 18 50 01

# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work? Yes, change of response for conflicts with existing booking or creation date.
- [ ] Have they been released to production already? No

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

### CODE REVIEW

File: server/testutils/factories/newRoom.ts [hunk 1]
```diff
@@ -21,5 +22,5 @@ export default NewRoomFactory.define(() => ({
     characteristic => characteristic.id,
   ),
   notes: faker.lorem.lines(),
-  bedEndDate: DateFormats.dateObjToIsoDate(faker.date.future({ years: 1 })),
+  bedEndDate: DateFormats.dateObjToIsoDate(addYears(faker.date.future({ years: 1 }), 2)),
```
Comment (richpjames) [inline server/testutils/factories/newRoom.ts:25]: why not just do `faker.date.future({ years: 3 })`?

**Review by richpjames — APPROVED** [review]
LGTM

File: server/testutils/factories/newRoom.ts (hunk 1 above)
Comment (froddd) [reply server/testutils/factories/newRoom.ts:25]: We want to ensure the date is _at least_ 2 years in the future -- `faker.date.future()` returns a future date _within_ the `years` given as options, as far as I know.

Comment (richpjames) [reply server/testutils/factories/newRoom.ts:25]: Makes sense, I knew there would be a good reason

**Pull request merged**